redhat-marketplace/redhat-marketplace-operator
# Feature/monitoring stack

Tests
- Add the Health conditions to the e2e test asserts (debt from another pr)
- Add UserWorkloadMonioting conditions to e2e test asserts
- Add userworkloadmonitoring-test e2e test set, described in tests/v2/e2e/userworkloadmonitoring-test/userworkloadmonitoring.md
- Modify meterbase_controller_test.go so it can discern RHOS version and handle lack of rhm-meterbase-additional-scrape-configs on 4.6

API Spec
- Add UserWorkloadMonitoringEnabled flag to MeterBase spec, add condition type,reason,messages for tracking status

Assets
- Add kube-state-metrics, kubelet, rhm-metric-state ServiceMonitors for UWM/4.6 prom instead of additional scrape configs
- Add a Secret that will be populated with redhat-marketplace-operator, such that there is a static named token that can be referenced for prometheus connection
- Add a prometheus deployment for 4.6 (prom v0.42.1). This is equivalent to UserWorkloadMonitoring and allows using ServiceMonitors with bearerTokenSecret references instead of hardcoded files
- Rename metric-state assets to distiguish 4.5 vs. 4.6

Controllers
- MeterBase 
  - Watches for changes to UserWorkloadMonioting ConfigMap assets
	- Handles UserWorkloadMonitoring and tracks the state in the status, including a 1d transitional
	- Installation of ServiceMonitors
	- Seperate out PrometheusServingCertsCABundle since it is also used by metric-state, in the UserWorkloadMonitoring case

- MeterDefinition
  - Relies on MeterBase instnace to determine which prometeus to query
	- Remove reconcile.Request from health funcs, it was never used

Configuration & assets
  - Handle using prom 4.6 or 4.5 images
	- Add semver parsing to infrastructure for version comparisons
  - Handle the new ServiceMonitors and versioned deployments in the factory
	- prevent a panic in getCredFromSecret if map access cameback nil
	
Prometheus
  - handle UserWorkloadMonitoring for prom api, client
	
Finalizer bug
  - With how finalizer result was being used, it should return Finished when successfully done, so the reconciler stops. Prior it would continue to install subcomponents, even after finalizer uninstalled things. Set to to consider Fininshed if the prior actions were success/Continue or already-deleted/OnNotFound. Deletes should be a lot snappier, since it is not churning now.
